### DISCUSSION TITLE
boolean_difference not working?

### DISCUSSION THREAD

**Discussion by bhaveshshrimali** [pyvista/pyvista#4905] [Q&A]
Essentially following [boolean_example](https://docs.pyvista.org/examples/01-filter/boolean-operations.html?highlight=boolean_difference) verbatim, save for a minor change

```python3
import pyvista as pv
pvplt = pv.Plotter()
pvplt.disable_depth_peeling()
s1 = pv.Sphere(radius=0.6, center=(0.5,0.5,0.5), theta_resolution=120, phi_resolution=120)
c1 = pv.Cube(center=(0.5, 0.5, 0.5), x_length=1, y_length=1, z_length=1)  #here
common_region = s1.boolean_difference(c1)
pvplt.add_mesh(common_region)
pvplt.view_isometric()
pvplt.show()
```

raises

```
RuntimeError: Empty meshes cannot be plotted. Input mesh has zero points.
```

but 

```python3
c1make = make_cube()
common_region = s1.boolean_difference(c1make)
pvplt.add_mesh(common_region, color="r")
```

plots without any problem.
  
![trialExample](https://user-images.githubusercontent.com/22624037/75850143-297c6e00-5dac-11ea-8d7f-b620e6879590.png)

Is this by design (that `Cube` wouldn't contain the enclosed volume) ? I am setting up `pymesh` following the reply on pyvista/pyvista-support#123 but wanted to try simple `booleans` with `pyvista`

**Comment by GuillaumeFavelier** [comment]
Hello @bhaveshshrimali ! I realized that the cells of `pv.Cube()` were actually quads so I managed to make your code snippet plot by using `triangulate()`:

```diff
- c1 = pv.Cube(center=(0.5, 0.5, 0.5), x_length=1, y_length=1, z_length=1)  #here
+ c1 = pv.Cube(center=(0.5, 0.5, 0.5), x_length=1, y_length=1, z_length=1).triangulate()
```

But still there is something here that I find counter-intuitive:

```py
import pyvista as pv
pvplt = pv.Plotter(shape=(1, 3))
pvplt.disable_depth_peeling()
s1 = pv.Sphere(radius=0.6, center=(0.5,0.5,0.5), theta_resolution=120, phi_resolution=120)
c1 = pv.Cube(center=(0.5, 0.5, 0.5), x_length=1, y_length=1, z_length=1).triangulate()
common_region = s1.boolean_difference(c1)

pvplt.subplot(0, 0)
pvplt.add_text("s1 + c1")
pvplt.add_mesh(s1 + c1)

pvplt.subplot(0, 1)
pvplt.add_text("s1 - c1")
pvplt.add_mesh(s1 - c1)

pvplt.subplot(0, 2)
pvplt.add_text("common region")
pvplt.add_mesh(common_region)

pvplt.view_isometric()
pvplt.show()
```

![2020-03-10_1920x1080](https://user-images.githubusercontent.com/18143289/76316939-9ecfcd80-62db-11ea-9f4f-3d6cb2c5e4d3.png)
